File: app/src/main/java/com/sdchang/permissionpolice/db/Cryo.java
package com.sdchang.permissionpolice.db;

import android.support.v4.util.Pools;
import com.esotericsoftware.kryo.Kryo;
import com.esotericsoftware.kryo.io.Input;
import com.esotericsoftware.kryo.io.Output;
import timber.log.Timber;

import javax.inject.Inject;

/**
 *
 */
public class Cryo {
    public static final int BUF_SIZE = 8192;
    public static final int BUF_MAX_SIZE = -1;
    private static final byte[] EMPTY_BUF = new byte[0];

    private Kryo mKryo;
    private Pools.Pool<Output> mOutputPool;
    private Pools.Pool<Input> mInputPool;

    @Inject
    public Cryo(Kryo kryo) {
        mKryo = kryo;
        mOutputPool = new Pools.SynchronizedPool<>(5);
        mInputPool = new Pools.SynchronizedPool<>(5);
    }

    public byte[] serialize(Object val) {
        registerType(val.getClass());
        Output output = acquireOutput();
        mKryo.writeObject(output, val);
        byte[] bytes = output.toBytes();
        release(output);
        return bytes;
    }

    public <T> T deserialize(byte[] data, Class<T> type) {
        registerType(type);
        Input input = acquireInput();
        input.setBuffer(data);
        T val = mKryo.readObject(input, type);
        release(input);
        return val;
    }

    public <T> T deserialize(byte[] data) {
        Input input = acquireInput();
        input.setBuffer(data);
        Class type = mKryo.readClass(input).getType();
        T val = null;
        try {
            val = (T) mKryo.readObject(input, type);
        } catch (ClassCastException e) {
            Timber.wtf(e, "Actual type: " + type.getCanonicalName());
        }
        release(input);
        return val;
    }

    private Output acquireOutput() {
        Output instance = mOutputPool.acquire();
        return (instance != null) ? instance : new Output(BUF_SIZE, BUF_MAX_SIZE);
    }

    private void release(Output instance) {
        instance.clear();
        mOutputPool.release(instance);
    }

    private Input acquireInput() {
        Input instance = mInputPool.acquire();
        return (instance != null) ? instance : new Input();
    }

    private void release(Input instance) {
        instance.setBuffer(EMPTY_BUF);
        mInputPool.release(instance);
    }

    private void registerType(Class type) {
        mKryo.register(type);
        if (type.isArray()) { // register component type if value is an array
            mKryo.register(type.getComponentType());
        }
    }
}

File: app/src/main/kotlin/com/permissionnanny/data/Cryo.kt
package com.permissionnanny.data

import android.support.v4.util.Pools
import com.esotericsoftware.kryo.Kryo
import com.esotericsoftware.kryo.io.Input
import com.esotericsoftware.kryo.io.Output
import javax.inject.Inject

/**

 */
class Cryo
@Inject
constructor(private val kryo: Kryo) {
    private val outputPool: Pools.Pool<Output>
    private val inputPool: Pools.Pool<Input>

    init {
        outputPool = Pools.SynchronizedPool<Output>(5)
        inputPool = Pools.SynchronizedPool<Input>(5)
    }

    fun serialize(value: Any): ByteArray {
        registerType(value.javaClass)
        val output = acquireOutput()
        kryo.writeObject(output, value)
        val bytes = output.toBytes()
        release(output)
        return bytes
    }

    fun <T> deserialize(data: ByteArray, type: Class<T>): T? {
        if (!isDataValid(data)) {
            return null
        }
        registerType(type)
        val input = acquireInput()
        input.buffer = data
        val value = kryo.readObject(input, type)
        release(input)
        return value
    }

    fun <T> deserialize(data: ByteArray): T? {
        if (!isDataValid(data)) {
            return null
        }
        val input = acquireInput()
        input.buffer = data
        val type = kryo.readClass(input).type
        var value: T? = null
        try {
            value = kryo.readObject(input, type) as T
        } catch (e: ClassCastException) {
            //            Timber.wtf(e, "Actual type: " + type.getCanonicalName());
        }

        release(input)
        return value
    }

    private fun isDataValid(data: ByteArray?): Boolean {
        return data != null && data.size > 0
    }

    private fun acquireOutput(): Output {
        val instance = outputPool.acquire()
        return instance ?: Output(BUF_SIZE, BUF_MAX_SIZE)
    }

    private fun release(instance: Output) {
        instance.clear()
        outputPool.release(instance)
    }

    private fun acquireInput(): Input {
        val instance = inputPool.acquire()
        return instance ?: Input()
    }

    private fun release(instance: Input) {
        instance.buffer = EMPTY_BUF
        inputPool.release(instance)
    }

    private fun registerType(type: Class<*>) {
        kryo.register(type)
        if (type.isArray) { // register component type if value is an array
            kryo.register(type.componentType)
        }
    }

    companion object {

        val BUF_SIZE = 8192
        val BUF_MAX_SIZE = -1
        private val EMPTY_BUF = ByteArray(0)
    }
}
